examples/gan/dcgan_tf_keras/dc_gan.py

```diff
@@ -0,0 +1,127 @@
+
+    model.add(layers.Conv2DTranspose(
+        1, (5, 5), strides=(2, 2), padding='same', use_bias=False, activation='tanh'
+    ))
+    assert model.output_shape == (None, 28, 28, 1)
+
+    return model
+
+
+def generator_loss(fake_output):
+    cross_entropy = tf.keras.losses.BinaryCrossentropy(from_logits=True)
+    return cross_entropy(tf.ones_like(fake_output), fake_output)
+
+
+def make_discriminator_model():
+    model = tf.keras.Sequential()
+    model.add(layers.Conv2D(
+        64, (5, 5), strides=(2, 2), padding='same', input_shape=[28, 28, 1]
+    ))
+    model.add(layers.LeakyReLU())
+    model.add(layers.Dropout(0.3))
+
+    model.add(layers.Conv2D(
+        128, (5, 5), strides=(2, 2), padding='same'
+    ))
+    model.add(layers.LeakyReLU())
+    model.add(layers.Dropout(0.3))
+
+    model.add(layers.Flatten())
+    model.add(layers.Dense(1))
+
+    return model
+
+
+def discriminator_loss(real_output, fake_output):
+    cross_entropy = tf.keras.losses.BinaryCrossentropy(from_logits=True)
+    real_loss = cross_entropy(tf.ones_like(real_output), real_output)
+    fake_loss = cross_entropy(tf.zeros_like(fake_output), fake_output)
+    total_loss = real_loss + fake_loss
+    return total_loss
+
+
+class DCGan(tf.keras.Model):
+    def __init__(self, batch_size, noise_dim):
+        super(DCGan, self).__init__()
+
+        self.generator = make_generator_model(noise_dim)
+        self.generator_loss = generator_loss
+
+        self.discriminator = make_discriminator_model()
+        self.discriminator_loss = discriminator_loss
+
+        self.batch_size = batch_size
+        self.noise_dim = noise_dim
+
+    def compile(self, discriminator_optimizer, generator_optimizer):
+        super(DCGan, self).compile()
+        self.discriminator_optimizer = discriminator_optimizer
+        self.generator_optimizer = generator_optimizer
+
+    def call(self, inputs, training=None, mask=None):
+        pass
+
+    def train_step(self, images):
+        noise = tf.random.normal([self.batch_size, self.noise_dim])
+
+        with tf.GradientTape() as gen_tape, tf.GradientTape() as disc_tape:
+            generated_images = self.generator(noise, training=True)
+
+            real_output = self.discriminator(images, training=True)
+            fake_output = self.discriminator(generated_images, training=True)
+
+            gen_loss = self.generator_loss(fake_output)
+            disc_loss = self.discriminator_loss(real_output, fake_output)
+
+        gradients_of_generator = gen_tape.gradient(
+            gen_loss, self.generator.trainable_variables)
+        gradients_of_discriminator = disc_tape.gradient(
+            disc_loss, self.discriminator.trainable_variables)
+
+        self.generator_optimizer.apply_gradients(
+            zip(gradients_of_generator, self.generator.trainable_variables))
+        self.discriminator_optimizer.apply_gradients(
+            zip(gradients_of_discriminator, self.discriminator.trainable_variables))
```

Comment:
Please apply `black` to this file.

(I thought I made a comment that said this already, but I can't find it in the github ui, so forgive me if this a repeat)